File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def solution(n, computers):
    answer = 0
    arr = [[] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if computers[i][j] == 1 and i != j:
                arr[i].append(j)

    visited = [0] * n

    def bfs(num):
        q = deque()
        q.append(num)
        visited[num] = 1

        while q:
            y = q.popleft()
            for x in arr[y]:
                if not visited[x]:
                    visited[x] = 1
                    q.append(x)

    for i in range(n):
        if not visited[i]:
            bfs(i)
            answer += 1

    for i in range(n):
        if not visited[i]:
            answer += 1
    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    
    static boolean[] visited;
    static ArrayList<ArrayList<Integer>> arr;
    
    public int solution(int n, int[][] computers) {
        visited = new boolean[n];
        arr = new ArrayList<>();
        for (int i = 0; i<n; i++) {
            arr.add(new ArrayList<>());
        }
        for (int i = 0; i<n; i++) {
            for (int j = 0; j<n; j++) {
                if (i!=j && computers[i][j]==1) {
                    arr.get(i).add(j);
                }
            }
        }
        int answer = 0;
        for (int i = 0; i<n; i++) {
            if (!visited[i]) {
                bfs(i);
                answer++;
            }
        }
        return answer;
    }
    
    static void bfs(int num) {
        Queue<Integer> q = new LinkedList<>();
        q.add(num);
        visited[num] = true;
        
        while (!q.isEmpty()) {
            int curr = q.poll();
            for (int i: arr.get(curr)) {
                if (!visited[i]) {
                    visited[i] = true;
                    q.add(i);
                }
            }
        }
    }
}